### DISCUSSION TITLE
V2 App Store Events Data handling

### DISCUSSION THREAD

**Discussion by abdulblueskylabs** [imdhemy/laravel-in-app-purchases#163] [Q&A] [answered]
How can we access data coming from app store v2? for example on DID RENEW, I need new expiry date.

**Comment by imdhemy** [comment, marked as answer]
You can find this documented here. 

https://imdhemy.com/laravel-iap-docs/docs/get-started/event-listeners

**Reply by abdulblueskylabs** [reply]
what about the other data like auto_renew_status_change_date_ms. I am trying to handle when customer cancel the subscription for next cycle. Want to record the action date and time. Is there any way for app_store?

**Reply by abdulblueskylabs** [reply]
I can see **SignedTransactionInfo** and **SignedRenewalInfo**  when I hit $event->getServerNotification()->getSubscription()->getProviderRepresentation(); but both informations  are encrypted. do I need to manually decrypt or is there any method in the package?

**Reply by imdhemy** [reply]
You don't need to decrypt them, you can get them as follows:

```php
 /** @var V2DecodedPayload $payload */
$payload = $event->getServerNotification()->getSubscription()->getProviderRepresentation();

$transactionInfo = $payload->getTransactionInfo();
$renewalInfo = $payload->getRenewalInfo();
```

BTW: It would be great if you submitted a PR for this issue. https://github.com/imdhemy/laravel-in-app-purchases/issues/164

-- EDIT --
Further explanation
https://github.com/imdhemy/laravel-in-app-purchases/discussions/181#discussioncomment-4033627

**Reply by abdulblueskylabs** [reply]
I can't access data using the above methods.
Seems **$payload->getTransactionInfo();** is not available in library

**Reply by imdhemy** [reply]
Double check the returned object is an instance of `V2DecodedPayload` and you are using the latest version of the package.

https://github.com/imdhemy/appstore-iap/blob/e14b23f1c26097abcdb9483f4b9eba16f04cbdd1/src/ServerNotifications/V2DecodedPayload.php#L165-L178

**Reply by abdulblueskylabs** [reply]
I think I am including **(https://github.com/imdhemy/laravel-in-app-purchases)** but this implementation is available in **https://github.com/imdhemy/appstore-iap**

**Reply by imdhemy** [reply]
App store IAP is included in LIAP.

https://github.com/imdhemy/laravel-in-app-purchases/blob/45b0d44897702018446bcba7cad2632269d96b85/composer.json#L19-L25

---
**What is the output of:**

```diff
/** @var V2DecodedPayload $payload */
$payload = $event->getServerNotification()->getSubscription()->getProviderRepresentation();

+ dd(get_class($payload));
```
